### DISCUSSION TITLE
Some potential landmarks filtered out during localizer's feature matching?

### DISCUSSION THREAD

**Discussion by Nikopikonano** [openMVG/openMVG#2025] [Q&A] [answered]
Hello openMVG community,

I have been staring openMVGs code related to camera localization. I was wondering if the descriptor matching may sometimes filter out good landmarks from the resection as the matching uses nearest neighbor distance ration filtering for the features? So, the second best match where the distance ratio is checked could be very close to the best as they both may present the same landmark from different views.

What do you think? Have I understood the code properly?

**Comment by pmoulon** [comment, marked as answer]
Hello @Nikopikonano ,
You are right we are using MatchDistanceRatio and found it effective most of the time (here https://github.com/openMVG/openMVG/blob/develop/src/openMVG/sfm/pipelines/localization/SfM_Localizer_Single_3DTrackObservation_Database.cpp#L94)

You could try to add an option to add a direct matching (1o1) and see how it performs. I think we could have risk to have some False Positive, but feel free to try and let us know how it goes.

**Reply by Nikopikonano** [reply]
Hey @pmoulon,

Thank you for confirming my understanding. I tried a trick to match the locating image features to only one descriptor per landmark and this seems to speed up the MatchDistanceRatio by ~40% (just a small test with one image!). Also few more correspondences were found for refinement.

**Reply by rjanvier** [reply]
Hello Niko. I personally use median of SIFT desc for each tracks to mitigate this phenomena.

**Reply by pmoulon** [reply]
@Nikopikonano if you want do a PR with some code and testing it would be nice. We could think about adding different retrieval modes for the image localizer.

**Reply by Nikopikonano** [reply]
@rjanvier  Thanks for the tip. Using the median feature sounds good idea.
@pmoulon I may do that later if I got more time for this. I think would need to test this without my other messy modifications in the code. @rjanvier if you have some clean code ready for this I don't mind if you do the PR.

**Reply by rjanvier** [reply]
Yes, it's a good idea for relatively short tracks (̃ 5) depending on our scene. As I mainly work on flat textured scene with low viewpoint variation it works very well in my cases. Unfortunately I do not have high quality code I can push to the public repo of openMVG for now, but it's easy to hack into the current code base.
